FileStorage << cv::Matx not supported

##### System information (version)
- OpenCV => 3.1
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2013

##### Detailed description

The FileStorage class does not support writing of cv::Matx objects.

##### Steps to reproduce

    #include <stdlib.h>
    #include <opencv2/core/core.hpp>
    void main(int argc, char **argv)
    {
        cv::FileStorage fs("dummy.yml", cv::FileStorage::MEMORY);
        fs << "a" << cv::Matx33f::eye();
    }


does not compile because of ``error C2665: 'cv::write' : none of the 10 overloads could convert all the argument types``:

    1>E:\DEV\3rdparty.git\OpenCV\3.1.0\include\opencv2/core/persistence.hpp(1065): error C2665: 'cv::write' : none of the 10 overloads could convert all the argument types
    1>          E:\DEV\3rdparty.git\OpenCV\3.1.0\include\opencv2/core/persistence.hpp(974): could be 'void cv::write(cv::FileStorage &,const cv::String &,const cv::Range &)'
    1>          E:\DEV\3rdparty.git\OpenCV\3.1.0\include\opencv2/core/persistence.hpp(663): or       'void cv::write(cv::FileStorage &,const cv::String &,const std::vector<cv::DMatch,std::allocator<_Ty>> &)'
    1>          with
    1>          [
    1>              _Ty=cv::DMatch
    1>          ]
    1>          E:\DEV\3rdparty.git\OpenCV\3.1.0\include\opencv2/core/persistence.hpp(662): or       'void cv::write(cv::FileStorage &,const cv::String &,const std::vector<cv::KeyPoint,std::allocator<_Ty>> &)'
    1>          with
    1>          [
    1>              _Ty=cv::KeyPoint
    1>          ]
    1>          E:\DEV\3rdparty.git\OpenCV\3.1.0\include\opencv2/core/persistence.hpp(661): or       'void cv::write(cv::FileStorage &,const cv::String &,const cv::SparseMat &)'
    1>          E:\DEV\3rdparty.git\OpenCV\3.1.0\include\opencv2/core/persistence.hpp(660): or       'void cv::write(cv::FileStorage &,const cv::String &,const cv::Mat &)'
    1>          E:\DEV\3rdparty.git\OpenCV\3.1.0\include\opencv2/core/persistence.hpp(659): or       'void cv::write(cv::FileStorage &,const cv::String &,const cv::String &)'
    1>          E:\DEV\3rdparty.git\OpenCV\3.1.0\include\opencv2/core/persistence.hpp(658): or       'void cv::write(cv::FileStorage &,const cv::String &,double)'
    1>          E:\DEV\3rdparty.git\OpenCV\3.1.0\include\opencv2/core/persistence.hpp(657): or       'void cv::write(cv::FileStorage &,const cv::String &,float)'
    1>          E:\DEV\3rdparty.git\OpenCV\3.1.0\include\opencv2/core/persistence.hpp(656): or       'void cv::write(cv::FileStorage &,const cv::String &,int)'
    1>          while trying to match the argument list '(cv::FileStorage, cv::String, const cv::Matx<float,3,3>)'
    1>          OpenCVExperiments.cpp(46) : see reference to function template instantiation 'cv::FileStorage &cv::operator <<<cv::Matx<float,3,3>>(cv::FileStorage &,const _Tp &)' being compiled
    1>          with
    1>          [
    1>              _Tp=cv::Matx<float,3,3>
    1>          ]

--- a/modules/core/include/opencv2/core/persistence.hpp
+++ b/modules/core/include/opencv2/core/persistence.hpp
@@ -777,7 +777,7 @@ template<typename _Tp, int m, int n> static inline void read(const FileNode& nod
     Mat temp;
     read(node, temp); // read as a Mat class
 
-    if ( !temp.data || temp.dims > 2 || temp.rows != m || temp.cols != n || temp.channels() != 1 )
+    if (temp.empty())
         value = default_matx;
     else
         value = Matx<_Tp, m, n>(temp);

--- a/modules/core/include/opencv2/core/persistence.hpp
+++ b/modules/core/include/opencv2/core/persistence.hpp
@@ -772,6 +772,17 @@ template<typename _Tp, int cn> static inline void read(const FileNode& node, Vec
     value = temp.size() != cn ? default_value : Vec<_Tp, cn>(&temp[0]);
 }
 
+template<typename _Tp, int m, int n> static inline void read(const FileNode& node, Matx<_Tp, m, n>& value, const Matx<_Tp, m, n>& default_matx = Matx<_Tp, m, n>())
+{
+    Mat temp;
+    read(node, temp); // read as a Mat class
+
+    if ( !temp.data || temp.dims > 2 || temp.rows != m || temp.cols != n || temp.channels() != 1 )
+        value = default_matx;
+    else
+        value = Matx<_Tp, m, n>(temp);
+}
+
 template<typename _Tp> static inline void read(const FileNode& node, Scalar_<_Tp>& value, const Scalar_<_Tp>& default_value)
 {
     std::vector<_Tp> temp; FileNodeIterator it = node.begin(); it >> temp;
@@ -952,6 +963,12 @@ void write(FileStorage& fs, const Vec<_Tp, cn>& v )
         write(fs, v.val[i]);
 }
 
+template<typename _Tp, int m, int n> static inline
+void write(FileStorage& fs, const Matx<_Tp, m, n>& x )
+{
+    write(fs, Mat(x)); // write as a Mat class
+}
+
 template<typename _Tp> static inline
 void write(FileStorage& fs, const Scalar_<_Tp>& s )
 {
@@ -1015,6 +1032,12 @@ void write(FileStorage& fs, const String& name, const Vec<_Tp, cn>& v )
 {
     cv::internal::WriteStructContext ws(fs, name, FileNode::SEQ+FileNode::FLOW);
     write(fs, v);
+}
+
+template<typename _Tp, int m, int n> static inline
+void write(FileStorage& fs, const String& name, const Matx<_Tp, m, n>& x )
+{
+    write(fs, name, Mat(x)); // write as a Mat class
 }
 
 template<typename _Tp> static inline

--- a/modules/core/test/test_io.cpp
+++ b/modules/core/test/test_io.cpp
@@ -1095,33 +1095,55 @@ TEST(Core_InputOutput, filestorage_yaml_advanvced_type_heading)
 TEST(Core_InputOutput, filestorage_matx_io)
 {
     Matx33d matxTest(1.234, 2, 3, 4, 5, 6, 7, 8, 9.876);
+
+    FileStorage writer("", FileStorage::WRITE | FileStorage::MEMORY);
+    writer << "matxTest" << matxTest;
+    String content = writer.releaseAndGetString();
+
+    FileStorage reader(content, FileStorage::READ | FileStorage::MEMORY);
+    Matx33d matxTestRead;
+    reader["matxTest"] >> matxTestRead;
+    ASSERT_TRUE( cv::norm(matxTest, matxTestRead, NORM_INF) == 0 );
+
+    reader.release();
+}
+
+TEST(Core_InputOutput, filestorage_matx_io_size_mismatch)
+{
     Matx32d matxTestWrongSize(1, 2, 3, 4, 5, 6);
+
+    FileStorage writer("", FileStorage::WRITE | FileStorage::MEMORY);
+    writer << "matxTestWrongSize" << matxTestWrongSize;
+    String content = writer.releaseAndGetString();
+
+    FileStorage reader(content, FileStorage::READ | FileStorage::MEMORY);
+    Matx33d matxTestRead;
+    try
+    {
+        reader["matxTestWrongSize"] >> matxTestRead;
+        FAIL() << "wrong size matrix read but no exception thrown";
+    }
+    catch (const std::exception&)
+    {
+    }
+
+    reader.release();
+}
+
+TEST(Core_InputOutput, filestorage_matx_io_with_mat)
+{
     Mat normalMat = Mat::eye(3, 3, CV_64F);
 
-    std::vector<FileStorage::Mode> formats;
-    formats.push_back(FileStorage::Mode::FORMAT_JSON);
-    formats.push_back(FileStorage::Mode::FORMAT_XML);
-    formats.push_back(FileStorage::Mode::FORMAT_YAML);
+    FileStorage writer("", FileStorage::WRITE | FileStorage::MEMORY);
+    writer << "normalMat" << normalMat;
+    String content = writer.releaseAndGetString();
 
-    for(size_t i = 0; i < formats.size(); i++)
-    {
-        FileStorage writer("", FileStorage::WRITE | FileStorage::MEMORY | formats[i]);
-        writer << "matxTest" << matxTest;
-        writer << "matxTestWrongSize" << matxTestWrongSize;
-        writer << "normalMat" << normalMat;
-        String content = writer.releaseAndGetString();
+    FileStorage reader(content, FileStorage::READ | FileStorage::MEMORY);
+    Matx33d matxTestRead;
+    reader["normalMat"] >> matxTestRead;
+    ASSERT_TRUE( cv::norm(Mat::eye(3, 3, CV_64F), matxTestRead, NORM_INF) == 0 );
 
-        FileStorage reader(content, FileStorage::READ | FileStorage::MEMORY);
-        Matx33d matxTestRead;
-        reader["matxTest"] >> matxTestRead;
-        ASSERT_TRUE( cv::norm(matxTest, matxTestRead, NORM_INF) == 0 );
-        reader["matxTestWrongSize"] >> matxTestRead;
-        ASSERT_TRUE( cv::norm(Matx33d(), matxTestRead, NORM_INF) == 0 );
-        reader["normalMat"] >> matxTestRead;
-        ASSERT_TRUE( cv::norm(Mat::eye(3, 3, CV_64F), matxTestRead, NORM_INF) == 0 );
-
-        reader.release();
-    }
+    reader.release();
 }
 
 TEST(Core_InputOutput, filestorage_keypoints_vec_vec_io)

--- a/modules/core/test/test_io.cpp
+++ b/modules/core/test/test_io.cpp
@@ -1092,6 +1092,38 @@ TEST(Core_InputOutput, filestorage_yaml_advanvced_type_heading)
     ASSERT_EQ(cv::norm(inputMatrix, actualMatrix, NORM_INF), 0.);
 }
 
+TEST(Core_InputOutput, filestorage_matx_io)
+{
+    Matx33d matxTest(1.234, 2, 3, 4, 5, 6, 7, 8, 9.876);
+    Matx32d matxTestWrongSize(1, 2, 3, 4, 5, 6);
+    Mat normalMat = Mat::eye(3, 3, CV_64F);
+
+    std::vector<FileStorage::Mode> formats;
+    formats.push_back(FileStorage::Mode::FORMAT_JSON);
+    formats.push_back(FileStorage::Mode::FORMAT_XML);
+    formats.push_back(FileStorage::Mode::FORMAT_YAML);
+
+    for(size_t i = 0; i < formats.size(); i++)
+    {
+        FileStorage writer("", FileStorage::WRITE | FileStorage::MEMORY | formats[i]);
+        writer << "matxTest" << matxTest;
+        writer << "matxTestWrongSize" << matxTestWrongSize;
+        writer << "normalMat" << normalMat;
+        String content = writer.releaseAndGetString();
+
+        FileStorage reader(content, FileStorage::READ | FileStorage::MEMORY);
+        Matx33d matxTestRead;
+        reader["matxTest"] >> matxTestRead;
+        ASSERT_TRUE( cv::norm(matxTest, matxTestRead, NORM_INF) == 0 );
+        reader["matxTestWrongSize"] >> matxTestRead;
+        ASSERT_TRUE( cv::norm(Matx33d(), matxTestRead, NORM_INF) == 0 );
+        reader["normalMat"] >> matxTestRead;
+        ASSERT_TRUE( cv::norm(Mat::eye(3, 3, CV_64F), matxTestRead, NORM_INF) == 0 );
+
+        reader.release();
+    }
+}
+
 TEST(Core_InputOutput, filestorage_keypoints_vec_vec_io)
 {
     vector<vector<KeyPoint> > kptsVec;

